server-spi/src/main/java/org/keycloak/storage/group/GroupLookupProvider.java

```diff
@@ -42,10 +42,13 @@ public interface GroupLookupProvider {
      * @return GroupModel with the corresponding name.
      */
     default GroupModel getGroupByName(RealmModel realm, GroupModel parent, String name) {
-        return searchForGroupByNameStream(realm, name, true, null, null)
-                .filter(groupModel -> parent == null
-                        ? groupModel.getParent() == null
-                        : parent.getId().equals(groupModel.getParentId())).findFirst().orElse(null);
+        if (parent == null) {
+            return searchForGroupByNameStream(realm, name, true, null, null)
+                    .filter(groupModel -> groupModel.getName().equals(name)).findFirst().orElse(null);
```

Comment:
@mhajas that was my first expectation as well. But I removed the getGroupByName in the MapGroupProvider to test the default impl and the test was failing. If you read the javadoc 
`Returns the group hierarchy with the given string in name for the given realm.
 For a matching group node the parent group is fetched by id (with all children) and added to the result stream.
 This is done until the group node does not have a parent (root group)`

in this case you are getting for example a stream with

<name> (group1 (toplevel group))
<parent1><name> (group2 (child of parent))
<parent3><parent2><name> (group3 (child of parent2..))

and we are only interested in the toplevel groups that means we need to filter out the groups with fetched parent group.

Reply:
I see. Thanks for the explanation, I missed that. Do you think it would make sense to have the default implementation to use `realm.getTopLevelGroups`? 

```java
default GroupModel getGroupByName(RealmModel realm, GroupModel parent, String name) {
    return (parent == null ? realm.getTopLevelGroupsStream() : parent.getSubGroupsStream())
            .filter(groupModel -> groupModel.getName().equals(name)).findFirst().orElse(null);
}
```

Reply:
@mhajs that's a much better idea. If you have a lot of toplevel Groups it might be slower. But that is less confusing and that is the default impl. If someone reports it's slow it can implemented in the class more efficient.

Reply:
Exactly, thanks @bohmber